devito/data.py

```diff
@@ -122,11 +126,91 @@ def _alloc_C_libcall(self, size, ctype):
         alignment = self.lib.getpagesize()
         ret = self.lib.posix_memalign(ctypes.byref(c_pointer), alignment, c_bytesize)
         if ret == 0:
-            return c_pointer
+            return c_pointer, (c_pointer, )
         else:
-            return None
+            return None, None
 
-    def free(self, c_pointer, size):
+    def free(self, c_pointer):
+        self.lib.free(c_pointer)
+
+
+class GuardAllocator(MemoryAllocator):
+    """
+    Memory allocator based on ``posix`` functions. The allocated memory is
+    aligned to page boundaries.  Additionally, it allocates extra memory
+    before and after the data, and configures it so that an SEGV is thrown
+    immediately if an out-of-bounds access occurs.
+
+    Further, the remainder region of the last page (which cannot be
+    protected), is poisoned with NaNs.
+    """
+
+    is_Posix = True
+
+    @classmethod
+    def initialize(cls):
+        handle = find_library('c')
+        if handle is not None:
+            cls.lib = ctypes.CDLL(handle)
+
+    def __init__(self, padding=1048576):
+        self.padding = padding
+
+    def _alloc_C_libcall(self, size, ctype):
+        if not self.available():
+            raise RuntimeError("Couldn't find `libc`'s `posix_memalign` to "
+                               "allocate memory")
+
+        pagesize = self.lib.getpagesize()
+        assert self.padding % pagesize == 0
+
+        npages_pad = self.padding // pagesize
+        nbytes_user = size * ctypes.sizeof(ctype)
+        npages_user = (nbytes_user + pagesize - 1) // pagesize
+
+        npages_alloc = 2*npages_pad + npages_user
+
+        c_bytesize = ctypes.c_ulong(npages_alloc * pagesize)
+        c_pointer = ctypes.cast(ctypes.c_void_p(), ctypes.c_void_p)
+        alignment = self.lib.getpagesize()
+        ret = self.lib.posix_memalign(ctypes.byref(c_pointer), alignment, c_bytesize)
+        if ret != 0:
+            return None, None
+
+        # generate pointers to the left padding, the user data, and the right pad
+        padleft_pointer = c_pointer
+        c_pointer = ctypes.c_void_p(c_pointer.value + self.padding)
+        padright_pointer = ctypes.c_void_p(c_pointer.value + npages_user * pagesize)
+
+        # and set the permissions on the pad memory to 0 (no access)
+        # if these fail, don't worry about failing the entire allocation
+        c_padsize = ctypes.c_ulong(self.padding)
+        print("protect", padleft_pointer, padright_pointer)
```

Comment:
leftover, or use logger (perhaps print out only in high-verbosity mode)

Reply:
oops